rpc/v8/subscriptions.go

```diff
@@ -253,138 +259,79 @@ func (h *Handler) SubscribeTransactionStatus(ctx context.Context, txHash felt.Fe
 		return 0, jsonrpc.Err(jsonrpc.MethodNotFound, nil)
 	}
 
-	// If the error is transaction not found that means the transaction has not been submitted to the feeder gateway,
-	// therefore, we need to wait for a specified time and at regular interval check if the transaction has been found.
-	// If the transaction is found during the timout expiry, then we continue to keep track of its status otherwise the
-	// websocket connection is closed after the expiry.
-	curStatus, rpcErr := h.TransactionStatus(ctx, txHash)
-	if rpcErr != nil {
-		if rpcErr != rpccore.ErrTxnHashNotFound {
-			return 0, rpcErr
-		}
+	var lastStatus TxnStatus = 0
 
-		timeout := time.NewTimer(subscribeTxStatusTimeout)
-		ticker := time.NewTicker(subscribeTxStatusTickerDuration)
-
-	txNotFoundLoop:
-		for {
-			select {
-			case <-timeout.C:
-				ticker.Stop()
-				return 0, rpcErr
-			case <-ticker.C:
-				curStatus, rpcErr = h.TransactionStatus(ctx, txHash)
-				if rpcErr != nil {
-					if rpcErr != rpccore.ErrTxnHashNotFound {
-						return 0, rpcErr
-					}
-					continue
-				}
-				timeout.Stop()
-				break txNotFoundLoop
+	return h.subscribe(ctx, w, subscriber{
+		onStart: func(ctx context.Context, id uint64, sub *subscription, _ any) error {
+			return h.getInitialTxStatus(ctx, sub, id, txHash, &lastStatus)
+		},
+		onReorg: func(ctx context.Context, id uint64, sub *subscription, reorg *sync.ReorgBlockRange) error {
+			return sendReorg(w, reorg, id)
+		},
+		onNewHead: func(ctx context.Context, id uint64, sub *subscription, head *core.Block) error {
+			if lastStatus < TxnStatusAcceptedOnL2 {
+				return h.checkTxStatus(ctx, sub, id, txHash, &lastStatus)
 			}
-		}
-	}
+			return nil
+		},
+		onPending: func(ctx context.Context, id uint64, sub *subscription, pending *core.Block) error {
+			if lastStatus < TxnStatusAcceptedOnL2 {
+				return h.checkTxStatus(ctx, sub, id, txHash, &lastStatus)
+			}
+			return nil
+		},
+		onL1Head: func(ctx context.Context, id uint64, sub *subscription, l1Head *core.L1Head) error {
+			return h.checkTxStatus(ctx, sub, id, txHash, &lastStatus)
+		},
+	})
+}
 
-	id := h.idgen()
-	subscriptionCtx, subscriptionCtxCancel := context.WithCancel(ctx)
-	sub := &subscription{
-		cancel: subscriptionCtxCancel,
-		conn:   w,
+// If the error is transaction not found that means the transaction has not been submitted to the feeder gateway,
+// therefore, we need to wait for a specified time and at regular interval check if the transaction has been found.
+// If the transaction is found during the timout expiry, then we continue to keep track of its status otherwise the
+// websocket connection is closed after the expiry.
+func (h *Handler) getInitialTxStatus(ctx context.Context, sub *subscription, id uint64, txHash felt.Felt, lastStatus *TxnStatus) error {
```

Comment:
`txHash` should be passed as a reference, unless you intentionally didn't.

`*txStatus` should be passed by value

Reply:
Yes, I agree it is usually worth it to be "pure". Makes code easier to read in the long term. Thank you!